### ISSUE TITLE
API and GUI

### ISSUE THREAD

**Issue by gedw99** [dbnet-io/dbnet#10]
The Web GUI is OK, and i saw your tried out the fancy tauri  rust / wasm approach. Now its React, like everything.

I use GIOUI which is golang that compiles to wasm, or desktop or mobile.
its pretty good for these types of Use cases and so wanted to give it a try.

SO for GIOUI, I just need a golang types and some http transprt to do RPC. I dont even need REST or OPENAPI, cause its golang called golang.

--

The other way is good old HTMX with golang and go templates. Its pretty damn fast and easy.

**Comment by flarco** [comment]
@gedw99 not at all, it's great to hear another point of view. I settled on React because there are tons of libraries, resources and options. And I like typescript for ease of refactoring. I use [Prime React](https://www.primefaces.org/primereact-v7/#/) for this, cause I am bad at Styling and GUI design. Add advanced features while keeping a good UX on the UI is hard :).

For Golang stuff, I've recently started using [Wails](https://wails.io/) and really like it. It does all the auto-typing conversion as well into JS, server side function calling, similar to GIOUI. I will probably port dbnet over to that at some point, but it needs to be able to build as server only as well. 

Thanks for sharing your thoughts! I welcome them.

**Comment by gedw99** [comment]
Us devs are always arguing over which is better ...

I used wails too. Still prefer GIOUI or HTMX any day of the week :) 

I might whip up a GIOUI GUI. Its very quikc. Is there shared goang types i can use and a transport ?

**Comment by gedw99** [comment]
maybe https://github.com/flarco/dbio/tree/main/iop ?

seems to have the bits i need.

**Comment by flarco** [comment]
Curious why prefer GIOUI? 

What do you mean by shared golang type? like the structs?

Also not understanding `a transport`? If you look at https://github.com/dbnet-io/dbnet/blob/master/server/server.go, this is where the endpoint comes from. Which uses https://github.com/dbrest-io/dbREST.

**Comment by gedw99** [comment]
why GIOUI ? thats a really hard one to justify... Its so subjective.

after 15 years of the crazy abstractions over HTML and AJAX, we have React etc etc etc.  Its why HTMX has become so popular and really good.
HTMX is crazy simple.

GIOUI is the exact opposite and does not care if its a browser, TV, mobile or desktop. It just wants a GPU canvas and off it goes.

so they are kind of the polar opposite BUT the common thing with both is i am just calling golang. 

you can go run . this:
https://github.com/gioui/gio-example/tree/main/component

if your on linux you might need some extra bits. I am on mac.
https://gioui.org/doc/install

Its very easy to extend from here to build very complex apps.

---

HTMX is also my fav.

this one is pretty nice: https://github.com/joerdav/go-htmx-examples

**Comment by gedw99** [comment]
Or this one !

https://github.com/emad-elsaid/debugger

its a Golang debugger gui written in golang to debug Golang.

Its a bit slow because of what it has to do but the layout possibilities are nice.

**Comment by gedw99** [comment]
> What do you mean by shared golang type? like the structs?
> 
> Also not understanding `a transport`? If you look at https://github.com/dbnet-io/dbnet/blob/master/server/server.go, this is where the endpoint comes from. Which uses https://github.com/dbrest-io/dbREST.

I am not sure yet. Yes i just need the types. I guess your transport is just HTTP, no push over websockets or SSE ?

I might just use NATS for the transport and router even.  That's eliminates a ton of layers and gives you global Load balancing, security, etc in one hit. Most devs utterly hate this :)

**Comment by flarco** [comment]
Correct, no push. I had started with websockets, but it was a bit crappy.
The backend has a timeout and returns status 202 when the query isn't done (see [here](https://github.com/dbrest-io/dbREST/blob/main/server/routes_query.go#L93) for backend logic).

See https://github.com/dbnet-io/dbnet/blob/master/frontend/src/state/dbnet.ts#L417-L440 for frontend logic, which is responsible for sending a new request.

I suggest you check out the following for backend abstraction, it has a cleaner code-base and none of the frontend stuff: 

- https://github.com/dbrest-io/dbREST/blob/main/server/routes_connection.go
- https://github.com/dbrest-io/dbREST/blob/main/server/routes_query.go
- https://github.com/dbrest-io/dbREST/blob/main/server/routes_table.go

dbREST is just the api part of dbNet. You could just call those functions by providing a fabricated `echo.Context`.

**Comment by flarco** [comment]
Thanks for all the links. I like `htmlx`'s simplicity alot. My gripe is that I would have to create all the html components myself (or grab from various places), and it would not be appealing to the eye (at least my eye :). Also I like typescript's typed nature, and IDE powered features. I'll def turn to `htmlx` for simpler projects. I also like GIOUI's simplicity, cool stuff.

Let me know how this NATS stuff goes (have not used that one before), I'm eager to see how this turns out. My goal for all this stuff was (is) to provide a mechanism to easily query databases of all kinds, as well as move data into/out of databases, hence the separation into 3 tools: 

<img width="776" alt="image" src="https://github.com/dbnet-io/dbnet/assets/7671010/4725133d-482c-4ce3-9d92-87c5d4a59a96">
screenshot from https://github.com/flarco

**Comment by gedw99** [comment]
sure will let you know. At least your not pisssed off :) 

I am a bit of a oldie and seen all the fads... so i admit my own bias.

I have gotten a GIO gui going with github.com/flarco/dbio. Early days... Will let you know and put it on github. Feel fee to copy it etc if you want.  If you like it then we can work on it together with your other packages.

GIO now does deep links and Url Scheme BTW. So you can compile for web or desktop and if the user has desktop it will open desktop and if not it will open in the browser.  

It also has webviews and other goodies: https://github.com/gioui-plugins/gio-plugins
this is why i dont use WAILS btw. GIo can bypass the Electron cruft. Maybe i am missing a trick but its pretty nice to use without the abstractions.

About  "creating all the html components yourself" - i totally get ya. With you get all the material design stuff for free with gio.   If you want a sexy one you can often find it and then just "go get " it.
